src/Integration.Vsix/SonarLintDaemon/Protocol.cs

```diff
@@ -0,0 +1,231 @@
+        public int Column { get; }
+        public int EndLine { get; }
+        public int EndColumn { get; }
+        public string Text { get; }
+
+        public MessagePart(
+      String filename,
+      int line,
+      int column,
+      int endLine,
+      int endColumn,
+      String text
+    )
+        {
+            Filename = filename;
+            Line = line;
+            Column = column;
+            EndLine = endLine;
+            EndColumn = endColumn;
+            Text = text;
+        }
+    }
+
+    class Message : MessagePart
+    {
+
+        public string RuleKey { get; }
+        public int RemediationCost { get; }
+        public bool PartsMakeFlow { get; }
+        public MessagePart[] Parts { get; }
+
+        public Message(
+      String ruleKey,
+      String filename,
+      int line,
+      int column,
+      int endLine,
+      int endColumn,
+      int remediationCost,
+      String message,
+      bool partsMakeFlow,
+      MessagePart[] parts
+    ) : base(filename, line, column, endLine, endColumn, message)
+        {
+            this.RuleKey = ruleKey;
+            this.RemediationCost = remediationCost;
+            this.PartsMakeFlow = partsMakeFlow;
+            this.Parts = parts;
+        }
+
+    }
+
+    class Protocol
+    {
+        /**
+          * This method does not close the provided stream.
+          */
+        public static void write(BinaryWriter writer, Request request)
+        {
+            writeUTF(writer, "IN");
+            write(writer, request.Options);
+            writer.Write(request.Flags);
+            writer.Write(request.MsVersion);
+            write(writer, request.IncludeDirs);
+            write(writer, request.FrameworkDirs);
+            write(writer, request.VfsOverlayFiles);
+            writeUTF(writer, request.ModuleName);
+            writeUTF(writer, request.Predefines);
+            write(writer, request.Macros);
+            writeUTF(writer, request.TargetTriple);
+            writeUTF(writer, request.File);
+            writeUTF(writer, "END");
+        }
+
+        private static void write(BinaryWriter writer, string[] strings)
+        {
+            writer.Write(strings.Length);
+            foreach (string value in strings)
+            {
+                writer.Write(value);
+            }
+        }
+
+        private static void writeUTF(BinaryWriter writer, string str)
+        {
+            // FIXME change to use modified UTF-8 encoding
+            writer.Write(str);
+        }
+        private static string readUTF(BinaryReader reader)
+        {
+            // FIXME change to use modified UTF-8 encoding
+            return reader.ReadString();
+        }
+
+        /**
+           * This method does not close the provided stream.
+           */
+        public static Response read(BinaryReader reader)
+        {
+
+            if ("OUT" != readUTF(reader))
+            {
+                throw new IOException();
```

Comment:
There's a more specific exception type [`[System.IO.InvalidDataException]`](https://docs.microsoft.com/en-us/dotnet/api/system.io.invaliddataexception?view=netframework-4.6.1) for data streams that "are in an invalid format".

It would be also be worth adding a message in the constructor as this will end up in the output window in SLVS e.g. "stream does not contain output data"